cranelift-filetests/src/function_runner.rs

```diff
@@ -1,53 +1,301 @@
+        macro_rules! gen_match {
+            ($x:ident, $($y: path => $z: ty, $t: path), +) => {
+                match $x {
+                    $($y => {
+                        if let $t(v) = expected {
+                            Self::check_assertion(Self::invoke::<$z>(&code_page), operator, v)
+                        } else {
+                            panic!()
+                        }
+                    }),+ ,
+                    _ => Err(format!("Unknown return type for check: {}", $x)),
+                };
+            };
+        }
+
+        gen_match!(ret_value_type,
+            types::I8 => u8, Value::I8,
+            types::I16 => u16, Value::I16,
+            types::I32 => u32, Value::I32,
+            types::I64 => u64, Value::I64,
+            types::I128 => u128, Value::I128,
+            types::B1 => bool, Value::Bool,
+            types::B8 => bool, Value::Bool,
+            types::B16 => bool, Value::Bool,
+            types::B32 => bool, Value::Bool,
+            types::B64 => bool, Value::Bool,
+            types::B128 => bool, Value::Bool,
+            types::F32 => f32, Value::F32,
+            types::F64 => f64, Value::F64
+        )
+    }
+
+    fn print(code_page: memmap::Mmap, ret_value_type: types::Type) -> Result<(), String> {
+        macro_rules! gen_match {
+            ($x:ident, $($y: path => $z: ty), +) => {
+                match $x {
+                    $($y => {
+                        Err(format!("Return value is: {}", Self::invoke::<$z>(&code_page)))
+                    }),+ ,
+                    _ => Err(format!("Unknown return type for check: {}", $x)),
+                };
+            };
+        }
+
+        gen_match!(ret_value_type,
+            types::I8 => u8,
+            types::I16 => u16,
+            types::I32 => u32,
+            types::I64 => u64,
+            types::B1 => bool,
+            types::B8 => bool,
+            types::B16 => bool,
+            types::B32 => bool,
+            types::B64 => bool,
+            types::B128 => bool,
+            types::F32 => f32,
+            types::F64 => f64
+        )
     }
 
     /// Compile and execute a single function, expecting a boolean to be returned; a 'true' value is
     /// interpreted as a successful test execution and mapped to Ok whereas a 'false' value is
     /// interpreted as a failed test and mapped to Err.
     pub fn run(&self) -> Result<(), String> {
         let func = self.function.clone();
-        if !(func.signature.params.is_empty()
-            && func.signature.returns.len() == 1
-            && func.signature.returns.first().unwrap().value_type.is_bool())
-        {
-            return Err(String::from(
-                "Functions must have a signature like: () -> boolean",
+
+        if !(func.signature.params.is_empty() && func.signature.returns.len() == 1) {
+            return Err(format!(
+                "Functions must have a signature like: () -> ty, found {}",
+                func.name
```

Comment:
The error message seems to be incorrect; maybe we can tell which function doesn't respect this, and what is its actual signature?

Reply:
Also, it would be worth merging this check in the `return_type` function, and call this function sooner, to avoid compiling if we can't check the result.